spec/app/domain/authentication/authn-azure/validate_application_identity_spec.rb

```diff
@@ -0,0 +1,433 @@
+          xms_mirid_token_field: xms_mirid_token_missing_initial_slash,
+          oid_token_field:       oid_token_field
+        )
+      end
+      it "does not raise an error" do
+        expect { subject }.to_not raise_error
+      end
+    end
+
+    context "with a valid application identity" do
+      context "and an Azure token with matching data" do
+        context "with no assigned Azure identity in application identity" do
+          subject do
+            Authentication::AuthnAzure::ValidateApplicationIdentity.new(
+              resource_class:             resource_class,
+              validate_azure_annotations: validate_azure_annotations,
+              ).call(
+              account:               account,
+              service_id:            service_id,
+              username:              username,
+              xms_mirid_token_field: xms_mirid_token_field_system_assigned_identity,
+              oid_token_field:       oid_token_field
+            )
+          end
+
+          it "does not raise an error" do
+            expect { subject }.to_not raise_error
+          end
+        end
+
+        context "with a user assigned Azure identity in application identity" do
+          before(:each) do
+            allow(host).to receive(:annotations)
+                             .and_return([subscription_id_annotation, resource_group_annotation, user_assigned_identity_annotation])
+          end
+          subject do
+            Authentication::AuthnAzure::ValidateApplicationIdentity.new(
+              resource_class:             resource_class,
+              validate_azure_annotations: validate_azure_annotations,
+              ).call(
+              account:               account,
+              service_id:            service_id,
+              username:              username,
+              xms_mirid_token_field: xms_mirid_token_field_user_assigned_identity,
+              oid_token_field:       oid_token_field
+            )
+          end
+
+          it "does not raise an error" do
+            expect { subject }.to_not raise_error
+          end
+        end
+
+        context "with a system assigned Azure identity in application identity" do
+          subject do
+            Authentication::AuthnAzure::ValidateApplicationIdentity.new(
+              resource_class:             resource_class,
+              validate_azure_annotations: validate_azure_annotations,
+              ).call(
+              account:               account,
+              service_id:            service_id,
+              username:              username,
+              xms_mirid_token_field: xms_mirid_token_field_system_assigned_identity,
+              oid_token_field:       oid_token_field
+            )
+          end
+
+          it "does not raise an error" do
+            expect { subject }.to_not raise_error
+          end
+        end
+      end
+
+      context "and a non-existent Role in Conjur" do
```

Comment:
If the role doesn't exist how is this context inside `"with a valid application identity"`? Can a non-existent role have a valid app id in its annotations?

Reply:
very good point